File: packages/go/config_test.go
package rdk

import (
	"strings"
	"testing"
)

func TestSettlementProofMatrix(t *testing.T) {
	cases := []struct {
		settlement SettlementParadigm
		proof      ProofSystem
		want       bool
	}{
		{SettlementOptimistic, ProofFraud, true},
		{SettlementOptimistic, ProofSnark, false},
		{SettlementZK, ProofSnark, true},
		{SettlementZK, ProofStark, true},
		{SettlementZK, ProofFraud, false},
		{SettlementBased, ProofNone, true},
		{SettlementBased, ProofSnark, false},
		{SettlementSovereign, ProofNone, true},
	}
	for _, c := range cases {
		if got := IsProofCompatible(c.settlement, c.proof); got != c.want {
			t.Errorf("IsProofCompatible(%s,%s)=%v want %v", c.settlement, c.proof, got, c.want)
		}
	}
	if !RequiresBasedSequencer(SettlementBased) {
		t.Error("based settlement should require based sequencer")
	}
	if RequiresBasedSequencer(SettlementZK) {
		t.Error("zk settlement should not require based sequencer")
	}
}

func TestValidateRollupConfig_Valid(t *testing.T) {
	cfg, err := PresetDefi().SetRollupID("my-rollup").Build()
	if err != nil {
		t.Fatalf("defi preset should build: %v", err)
	}
	r := ValidateRollupConfig(cfg)
	if !r.Valid || len(r.Errors) != 0 {
		t.Errorf("expected valid, got %+v", r)
	}
	if len(r.Warnings) != 0 {
		t.Errorf("defi (native DA) should have no warnings, got %v", r.Warnings)
	}
}

func TestValidateRollupConfig_MatrixError(t *testing.T) {
	cfg := RollupConfig{
		RollupID: "r", Profile: ProfileCustom,
		Settlement: SettlementZK, Sequencer: SequencerDedicated,
		DA: DANative, ProofSystem: ProofFraud, GasModel: GasStandard,
		VmType: VmEVM, BlockTimeMs: 1000, MaxTxPerBlock: 1000,
	}
	r := ValidateRollupConfig(cfg)
	if r.Valid {
		t.Fatal("zk + fraud should be invalid")
	}
	found := false
	for _, e := range r.Errors {
		if e == `proof system "fraud" is not compatible with "zk" settlement (expected one of: snark, stark)` {
			found = true
		}
	}
	if !found {
		t.Errorf("missing matrix error, got %v", r.Errors)
	}
}

func TestValidateRollupConfig_BasedSequencerRequired(t *testing.T) {
	cfg := RollupConfig{
		RollupID: "r", Profile: ProfileEnterprise,
		Settlement: SettlementBased, Sequencer: SequencerDedicated,
		DA: DANative, ProofSystem: ProofNone, GasModel: GasFlat,
		VmType: VmEVM, BlockTimeMs: 1000, MaxTxPerBlock: 1000,
	}
	r := ValidateRollupConfig(cfg)
	if r.Valid {
		t.Fatal("based settlement with dedicated sequencer should be invalid")
	}
}

func TestValidateRollupConfig_CelestiaWarning(t *testing.T) {
	cfg := PresetNFT().SetRollupID("nft-1").Get()
	r := ValidateRollupConfig(cfg)
	if !r.Valid {
		t.Fatalf("nft preset should be valid, got %v", r.Errors)
	}
	if len(r.Warnings) == 0 {
		t.Error("nft preset (celestia DA) should emit a warning")
	}
}

func TestValidateRollupConfig_FieldSanity(t *testing.T) {
	cfg := RollupConfig{
		RollupID: "", Settlement: SettlementOptimistic, Sequencer: SequencerDedicated,
		DA: DANative, ProofSystem: ProofFraud, GasModel: GasStandard, VmType: VmEVM,
		BlockTimeMs: 0, MaxTxPerBlock: -1, StakeAmountUqor: "abc",
	}
	r := ValidateRollupConfig(cfg)
	if r.Valid {
		t.Fatal("expected invalid")
	}
	wantSubstrings := []string{"rollupId", "blockTimeMs", "maxTxPerBlock", "stakeAmountUqor"}
	for _, sub := range wantSubstrings {
		found := false
		for _, e := range r.Errors {
			if strings.Contains(e, sub) {
				found = true
			}
		}
		if !found {
			t.Errorf("expected an error mentioning %q, got %v", sub, r.Errors)
		}
	}
}

func TestBuilderToCreateMsg(t *testing.T) {
	b := PresetDefi().SetRollupID("r1").SetStakeAmountUqor("10000000000")
	msg, err := b.ToCreateMsg("qor1creator", "")
	if err != nil {
		t.Fatal(err)
	}
	if msg.Creator != "qor1creator" || msg.RollupID != "r1" || msg.Profile != ProfileDefi ||
		msg.VmType != VmEVM || msg.StakeAmount != "10000000000" {
		t.Errorf("unexpected create msg: %+v", msg)
	}

	if _, err := PresetDefi().SetRollupID("r1").ToCreateMsg("qor1c", ""); err == nil {
		t.Error("expected error when no stake amount supplied")
	}
}

File: packages/jvm/src/test/java/io/github/qorechain/rdk/ConfigTest.java
package io.github.qorechain.rdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.github.qorechain.rdk.config.Enums.ProofSystem;
import io.github.qorechain.rdk.config.Enums.SequencerMode;
import io.github.qorechain.rdk.config.Enums.SettlementParadigm;
import io.github.qorechain.rdk.config.Matrix;
import io.github.qorechain.rdk.config.RollupConfig;
import io.github.qorechain.rdk.config.Validate;
import io.github.qorechain.rdk.config.ValidationResult;
import io.github.qorechain.rdk.presets.Presets;
import java.util.List;
import org.junit.jupiter.api.Test;

class ConfigTest {

    @Test
    void matrixMatchesChainRules() {
        assertEquals(List.of(ProofSystem.FRAUD), Matrix.validProofSystems(SettlementParadigm.OPTIMISTIC));
        assertEquals(
                List.of(ProofSystem.SNARK, ProofSystem.STARK),
                Matrix.validProofSystems(SettlementParadigm.ZK));
        assertEquals(List.of(ProofSystem.NONE), Matrix.validProofSystems(SettlementParadigm.BASED));
        assertEquals(List.of(ProofSystem.NONE), Matrix.validProofSystems(SettlementParadigm.SOVEREIGN));

        assertTrue(Matrix.isProofCompatible(SettlementParadigm.ZK, ProofSystem.SNARK));
        assertFalse(Matrix.isProofCompatible(SettlementParadigm.ZK, ProofSystem.FRAUD));
        assertTrue(Matrix.requiresBasedSequencer(SettlementParadigm.BASED));
        assertFalse(Matrix.requiresBasedSequencer(SettlementParadigm.OPTIMISTIC));
    }

    @Test
    void validateAcceptsAValidPreset() {
        RollupConfig config = Presets.defi(c -> c.rollupId = "ok").get();
        ValidationResult r = Validate.validateRollupConfig(config);
        assertTrue(r.valid, () -> "errors: " + r.errors);
    }

    @Test
    void validateRejectsAnIncompatibleProofPair() {
        RollupConfig config = Presets.defi(c -> c.rollupId = "bad").get();
        config.proofSystem = ProofSystem.FRAUD; // zk settlement requires snark|stark
        ValidationResult r = Validate.validateRollupConfig(config);
        assertFalse(r.valid);
        assertTrue(r.errors.stream().anyMatch(e -> e.contains("not compatible")));
    }

    @Test
    void validateEnforcesBasedSequencerForBasedSettlement() {
        RollupConfig config = Presets.gaming(c -> c.rollupId = "g").get();
        config.sequencer = SequencerMode.DEDICATED; // based settlement requires based sequencer
        ValidationResult r = Validate.validateRollupConfig(config);
        assertFalse(r.valid);
        assertTrue(r.errors.stream().anyMatch(e -> e.contains("based settlement requires")));
    }

    @Test
    void validateWarnsOnCelestia() {
        RollupConfig config = Presets.nft(c -> c.rollupId = "n").get();
        ValidationResult r = Validate.validateRollupConfig(config);
        assertTrue(r.valid, () -> "errors: " + r.errors);
        assertTrue(r.warnings.stream().anyMatch(w -> w.contains("Celestia")));
    }

    @Test
    void validateRejectsEmptyRollupId() {
        RollupConfig config = Presets.defi(null).get();
        ValidationResult r = Validate.validateRollupConfig(config);
        assertFalse(r.valid);
        assertTrue(r.errors.stream().anyMatch(e -> e.contains("rollupId")));
    }
}
